torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1599,9 +1599,31 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
     int deviceIndex = devices[i].index();
 
     gpuGuard.set_index(deviceIndex);
+#ifdef NCCL_HAS_COMM_SPLIT
+    if (options_->split_from) {
+      TORCH_CHECK(
+          options_->split_color != 0,
+          "Must specify a non-zero color when splitting");
+      // Find a valid, healthy communicator to split from if possible.
+      std::lock_guard<std::mutex> lock(options_->split_from->mutex_);
+      if (!options_->split_from->deviceComms_.empty() &&
+          options_->split_from->deviceComms_[i] &&
+          !options_->split_from->deviceComms_[i]->isAborted()) {
+        ncclComms[i] = NCCLComm::split(
+            options_->split_from->deviceComms_[i].get(),
+            options_->split_color,
+            rank,
+            options_->config);
+      }
+    }
+#endif
 #ifdef NCCL_HAS_COMM_NONBLOCKING
-    ncclComms[i] = NCCLComm::create(numRanks, rank, ncclID, options_->config);
+    if (!ncclComms[i]) {
+      ncclComms[i] = NCCLComm::create(numRanks, rank, ncclID, options_->config);
+    }
 #else
+    // ncclCommSplit was introduced after non-blocking, so for older
+    // NCCLs we fall back.
     ncclComms[i] = NCCLComm::create(numRanks, rank, ncclID);
 #endif
```

Comment:
I like the careful addition of `if (!ncclComms[i])` to cover the case where we cannot use split even though the api is available.

I wonder if the logic would be clearer `if (!ncclComms[i])` line is made the outer line, i.e.
```
if (!ncclComms[i]) {
#ifdef NCCL_HAS_COMM_NONBLOCKING
    ncclComms[i] = NCCLComm::create(numRanks, rank, ncclID, options_->config);
#else
    ncclComms[i] = NCCLComm::create(numRanks, rank, ncclID);
#endif
}
```